### PR TITLE
Replace Double-SHA256 with blake2b and implement domain seperation

### PR DESCRIPTION
**Pull request by elichai**
This solves the following issues: #1204, #1188, #1186.

We chose Blake2b because of its speeds (See https://github.com/kaspanet/kaspad/issues/1188#issue-757981948) we use blake2b's MAC for domain separation because golang x/crypto currently doesn't expose the `personalization` feature (See https://github.com/golang/go/issues/32447).

We also fix the merkle tree bug by doing 2 things:
1. Domain seperation for txs and for merkle branches.
2. If there's an odd number of leafs/branches instead of duplicating we append zeros. 

There's now a new file called `domains.go` that contains all the constructors for the domain separated writers. which should be the only way to hash in kaspad.
if we ever need to hash something new we need to write a new constructor with the appropriate domain

### CODE REVIEW

File: domain/consensus/utils/merkle/merkle.go [hunk 1]
```diff
@@ -90,12 +82,12 @@ func merkleRoot(hashes []*externalapi.DomainHash) *externalapi.DomainHash {
 			merkles[offset] = nil
 
 		// When there is no right child, the parent is generated by
-		// hashing the concatenation of the left child with itself.
+		// hashing the concatenation of the left child with zeros.
```
Comment (stasatdaglabs) [inline domain/consensus/utils/merkle/merkle.go:85]: What are the benefits of this over concatenating the child with itself?

File: domain/consensus/utils/txscript/opcode.go [hunk 2]
```diff
@@ -1962,17 +1960,17 @@ func opcodeHash160(op *parsedOpcode, vm *Engine) error {
...
-func opcodeHash256(op *parsedOpcode, vm *Engine) error {
+// Stack transformation: [... x1] -> [... blake2b(x1)]
+func opcodeBlake2b(op *parsedOpcode, vm *Engine) error {
 	buf, err := vm.dstack.PopByteArray()
 	if err != nil {
 		return err
 	}
-
-	vm.dstack.PushByteArray(hashes.HashData(buf)[:])
+	hash := sha256.Sum256(buf)
```
Comment (stasatdaglabs) [inline domain/consensus/utils/txscript/opcode.go:1972]: Is this correct?

**Review by stasatdaglabs — CHANGES_REQUESTED** [review]

File: domain/consensus/utils/merkle/merkle.go (hunk 1 above)
Comment (elichai) [reply domain/consensus/utils/merkle/merkle.go:85]: This prevents having 2 separate inclusion proofs for the same thing in the same merkle tree

Comment (elichai) [reply domain/consensus/utils/merkle/merkle.go:85]: And it is also the standard for how merkle trees are built, see: https://tools.ietf.org/html/rfc6962#section-2.1

File: domain/consensus/utils/txscript/opcode.go (hunk 2 above)
Comment (elichai) [reply domain/consensus/utils/txscript/opcode.go]: No :)

**Review by stasatdaglabs — APPROVED** [review]

**Pull request merged**